src/rust/engine/fs/store/src/lib.rs

```diff
@@ -1289,6 +1331,10 @@ impl Store {
           let path = destination.join(child.name().as_ref());
           let store = store.clone();
           child_futures.push(async move {
+            let can_be_immutable = immutable_inputs.is_some()
+              && !mutable_path_ancestors.contains(&path)
+              && !force_mutable;
+
             match child {
```

Comment:
This method body ended up pretty deeply nested. Maybe it would be worth breaking out a "materialize_child" method...? Unclear.